docs: Add Argo CD Core documentation (#13225)
argoproj/argo-cd: docs/operator-manual/core.md

# Argo CD Core

## Introduction

> Argo CD Core is mainly a different installation that makes Argo CD run
> in headless mode. With this installation you will have a fully

Reviewer: Nit: wording
```suggestion
Argo CD Core is a different installation method that runs Argo CD
in headless mode. With this installation, you will have a fully
```

> functional gitops engine capable of getting the desired state from Git

Author: Thank you! Fixed all occurrences in the doc with the suggested capitalization

applying it in Kubernetes.

The following groups of features won't be available in this
installation:

- Argo CD RBAC model
- Argo CD API
- OIDC based authentication
- Argo CD web application

> - Argo CD CLI

Reviewer 2: Is this completely true? I thought some CLI commands have flags which allow you to perform equivalent actions using the k8s API instead of the Argo CD API.

Author: Actually in the usage section below I am documenting how the CLI and UI can still be used. I created a new section for partially available features.

> - Multi-tenancy

Reviewer 2: You can still use Projects to implement a type of multi-tenancy. That multi-tenancy will simply be enforced by git push permissions (strictly gitops) instead of API RBAC.

Author: fair.. I created a new section providing a brief explanation.

A few use-cases that justify running Argo CD core are:

- As a cluster admin, I want to rely on Kubernetes RBAC only.
- As a devops engineer, I don't want to learn a new API or depend on
  another CLI to automate my deployments. I want instead rely in
  Kubernetes API only.
- As a cluster admin, I don't want to provide Argo CD UI or Argo CD
  CLI to developers.

## Architecture

Because Argo CD is designed with a component based architecture in
mind, it is possible to have a more minimalist installation. In this

> case fewer components are installed and yet the main gitops

Author: Thank you! Fixed all occurrences in the doc with the suggested capitalization.

functionality remains operational.

In the diagram below, the Core box, shows the components that will be
installed while opting for Argo CD Core:

![Argo CD Core](../assets/argocd-core-components.png)

Note that even if Argo CD controller can run without Redis, it isn't

> recommended. Redis is used by Argo CD controller as an important

Reviewer: Nit: wording
```suggestion
recommended. The Argo CD controller uses Redis as an important
```

caching mechanism reducing the load on Kube API and in Git. For this
reason, Redis is also included in this installation method.

## Installing

Argo CD Core can be installed by applying a single manifest file that
contains all the required resources.

Example:

```

> export ARGOCD_VERSION=<desired argo cd release version>

Reviewer 2: Maybe note that the version should start with `v` (provide an example).

Author: added

kubectl create namespace argocd
kubectl apply -n argocd -f https://raw.githubusercontent.com/argoproj/argo-cd/$ARGOCD_VERSION/manifests/core-install.yaml
```

## Using

Once Argo CD Core is installed, users will be able to interact with it
by relying on Gitops. The available Kubernetes resources will be the
Application and the ApplicationSet CRDs. By using those resources,
users will be able to deploy and manage applications in Kubernetes.

It is still possible to use Argo CD CLI even when running Argo CD
Core. In this case, the CLI will spawn a local API server process that
will be used to handle the CLI command. Once the command is concluded,
the local API Server process will also be terminated. This happens
transparently for the user with no additional command required. Note
that Argo CD Core will rely only on Kubernetes RBAC and the user (or
the process) invoking the CLI needs to have access to the Argo CD

> namespace with the proper permission in the Application and
> ApplicationSet resources for executing a given command.

Reviewer: Nit: formatting
```suggestion
namespace with the proper permission in the `Application` and
`ApplicationSet` resources for executing a given command.
```

> In order to use Argo CD CLI in core mode it is necessary to pass a
> special flag `--core` in the login command.

Reviewer: Nit: wording
```suggestion
To use Argo CD CLI in core mode, it is required to pass the `--core`
flag with the login command.
```
